### PR TITLE
MM-56355: Use same number of hubs as CPUs

### PR DESCRIPTION
**Pull request by agnivade**
We randomly used double the number of hubs. A proper
load test reveals a 14% improvement in number of users
by using the same number as CPUs.

Master: https://snapshots.raintank.io/dashboard/snapshot/mpDl0cJyycgCpA1J8mtTJGSrlO2AhN42
PR: https://snapshots.raintank.io/dashboard/snapshot/rFZPH2jHq1IuVSlJXuBI44zw1kV3VCwy
Half the number of hubs: https://snapshots.raintank.io/dashboard/snapshot/gchtzy2q4HUJ7V67LOX6Bzcy0V4MecGE

https://mattermost.atlassian.net/browse/MM-56355

```release-note
NONE
```

### CODE REVIEW

File: server/channels/app/platform/web_hub.go [hunk 1]
```diff
@@ -92,8 +92,7 @@ func newWebHub(ps *PlatformService) *Hub {
 
 // hubStart starts all the hubs.
 func (ps *PlatformService) hubStart(broadcastHooks map[string]BroadcastHook) {
-	// Total number of hubs is twice the number of CPUs.
-	numberOfHubs := runtime.NumCPU() * 2
+	numberOfHubs := runtime.NumCPU()
```
Comment (agarciamontoro) [inline server/channels/app/platform/web_hub.go:95]: Nit: maybe we should add a comment linking to this PR to explain why this number was chosen.

**Review by agarciamontoro — APPROVED** [review]
I'm wondering whether we should make this a setting that can be configured. I guess the optimal value will change depending on the number of CPUs, but for that we'd need a much more time-consuming analysis, so I'm ok hardcoding it for now. Maybe something for the future?

How many CPUs did your test setup have?

**Comment by agnivade** [comment]
Hmm .. why do you think optimal value will depend on number of CPUs? Intuitively, I'd think number of hubs = number of CPUs would make the most sense.

File: server/channels/app/platform/web_hub.go (hunk 1 above)
Comment (streamer45) [reply server/channels/app/platform/web_hub.go:95]: +1, let's keep the rationale documented this time, link to test results and discussion would be good.

**Review by streamer45 — COMMENTED** [review]
I was half joking when I mentioned it earlier, but I'd almost want to test this with less than `NumCPU`. I am generally wondering whether on top of thread switching there's an actual transfer of CPU time to a place that requires it more than the hub. If the theory is right then there could be more improvement in constraining the hub resources further.

**Comment by agnivade** [comment]
From looking at the CPU profiles, the hub is where 20% of the time is spent, so it's unlikely that there's a place that requires more CPU time than the hub. But for consistency's sake, I can run another test to be sure. :+1:

**Comment by agnivade** [comment]
Just ran a test with half the number of hubs: https://snapshots.raintank.io/dashboard/snapshot/gchtzy2q4HUJ7V67LOX6Bzcy0V4MecGE. Result seems to be in the same ballpark as equal hubs. 

I am thinking that it's clear that anything more than the number of CPUs is bad. It's just that we don't know the exact sweet spot yet.

To really dig deep into this, we need to come up with an exhaustive test strategy with varying instance sizes and number of hubs. Also, we need to devise a test that just stresses the websocket part of the system. Probably a modification to simple controller, to just send out typing messages. 

The question that I have in mind is whether we spend the effort now (it's probably a solid 1-2 weeks of effort), or merge the PR as-is, and revisit this when we have other parts of the system well-optimized as well (like the DB part).

**Comment by agnivade** [comment]
Ran a custom test and finally got some answers:

hubs

The test was done by writing a custom Go client which creates 10K clients and have one client send out typing messages once every 125ms. (8 per sec * 10000=80k msgs/sec).

Test1: double the number of hubs
Test2: equal hubs
Test3: half

Everything was done on my laptop. Each test was run for ~10mins.

We can see that all 3 tests can push the same number of messages, but they have differing performance characteristics.

**Test1** seems to be the worst off all because not only it takes higher CPU, but also the broadcast buffer gets filled up from time to time, leading to spikes.
**Test2** is much better with lower CPU usage and flatter broadcast buffer. This is because there are fewer hubs now, so the system has to spend less time serving each hub. Therefore, there are fewer spikes.
**Test3** has the best broadcast buffer usage, but uses higher CPU than test2.

I would say test2 and test3 are in the same ballpark. But it's clear that test1 is far worse. I'd say that using the same number of hubs is a safe change to go ahead with.

cc @streamer45 @agarciamontoro

**Review by streamer45 — APPROVED** [review]
Great work @agnivade , exactly what I wanted to see :raised_hands: Do you have a link to the code used to benchmark? Would be good to keep it around for reference.

**Comment by agnivade** [comment]
E2E tests look good. Merging!

**Pull request merged**